server/server.go

```diff
@@ -0,0 +1,825 @@
+	if _, ok := key.Public().(*rsa.PublicKey); !ok {
+		log.Errorf("Worker %v: %s: request is RSA, but key isn't\n", w.name, protocol.ErrCrypto)
+		w.s.stats.logInvalid(requestBegin)
+		return makeErrResponse(pkt.ID, protocol.ErrCrypto)
+	}
+
+	sig, err := key.Sign(rand.Reader, pkt.Operation.Payload, opts)
+	if err != nil {
+		log.Errorf("Worker %v: %s: Signing error: %v\n", w.name, protocol.ErrCrypto, err)
+		w.s.stats.logInvalid(requestBegin)
+		return makeErrResponse(pkt.ID, protocol.ErrCrypto)
+	}
+
+	w.s.stats.logRequestDuration(requestBegin)
+	return makeRespondResponse(pkt.ID, sig)
+}
+
+const randBufferLen = 1024 * 1024
+
+type ecdsaWorker struct {
+	buf  *buf_ecdsa.SyncRandBuffer
+	s    *Server
+	name string
+}
+
+func newECDSAWorker(s *Server, buf *buf_ecdsa.SyncRandBuffer, name string) *ecdsaWorker {
+	return &ecdsaWorker{
+		buf:  buf,
+		s:    s,
+		name: name,
+	}
+}
+
+func (w *ecdsaWorker) Do(job interface{}) interface{} {
+	pkt := job.(*protocol.Packet)
+
+	requestBegin := time.Now()
+	log.Debugf("Worker %v: version:%d.%d id:%d body:%s", w.name, pkt.MajorVers, pkt.MinorVers, pkt.ID, pkt.Operation)
+
+	var opts crypto.SignerOpts
+	var key crypto.Signer
+	var ok bool
+	switch pkt.Operation.Opcode {
+	case protocol.OpECDSASignMD5SHA1:
+		opts = crypto.MD5SHA1
+	case protocol.OpECDSASignSHA1:
+		opts = crypto.SHA1
+	case protocol.OpECDSASignSHA224:
+		opts = crypto.SHA224
+	case protocol.OpECDSASignSHA256:
+		opts = crypto.SHA256
+	case protocol.OpECDSASignSHA384:
+		opts = crypto.SHA384
+	case protocol.OpECDSASignSHA512:
+		opts = crypto.SHA512
+	default:
+		// It's the client's responsibility to send all non-ECDSA requests to the
+		// pool of otherWorkers.
+		panic(fmt.Sprintf("internal error: got unexpected opcode %v", pkt.Operation.Opcode))
+	}
+
+	if key, ok = w.s.keys.Get(&pkt.Operation); !ok {
+		log.Error(protocol.ErrKeyNotFound)
+		w.s.stats.logInvalid(requestBegin)
+		return makeErrResponse(pkt.ID, protocol.ErrKeyNotFound)
+	}
+
+	// Ensure we don't perform an RSA sign for an ECDSA request.
+	if _, ok := key.Public().(*ecdsa.PublicKey); !ok {
+		log.Errorf("Worker %v: %s: request is ECDSA, but key isn't\n", w.name, protocol.ErrCrypto)
+		w.s.stats.logInvalid(requestBegin)
+		return makeErrResponse(pkt.ID, protocol.ErrCrypto)
+	}
+
+	var sig []byte
+	var err error
+	if k, ok := key.(*ecdsa.PrivateKey); ok && k.Curve == elliptic.P256() {
+		sig, err = buf_ecdsa.Sign(rand.Reader, k, pkt.Operation.Payload, opts, w.buf)
+	} else {
+		sig, err = key.Sign(rand.Reader, pkt.Operation.Payload, opts)
+		log.Debugf("Worker %v: Computed ECDSA signature without buffer", w.name)
+	}
+	if err != nil {
+		log.Errorf("Worker %v: %s: Signing error: %v\n", w.name, protocol.ErrCrypto, err)
+		w.s.stats.logInvalid(requestBegin)
+		return makeErrResponse(pkt.ID, protocol.ErrCrypto)
+	}
+
+	w.s.stats.logRequestDuration(requestBegin)
+	return makeRespondResponse(pkt.ID, sig)
+}
+
+type backgroundWorker struct {
```

Comment:
I think`entropyWorker` is more appropriate. Every worker is background worker at some time.